projects/plugins/boost/app/modules/optimizations/lcp/class-lcp-optimize-bg-image.php

```diff
@@ -0,0 +1,212 @@
+		if ( LCP_Optimization_Util::should_skip_optimization() ) {
+			return;
+		}
+
+		$storage  = new LCP_Storage();
+		$lcp_data = $storage->get_current_request_lcp();
+
+		if ( empty( $lcp_data ) ) {
+			return;
+		}
+
+		$instance = new self( $lcp_data );
+
+		// Preload the background image as early as possible.
+		add_action( 'wp_head', array( $instance, 'preload_background_images' ), 1 );
+
+		// Add the background image styling as late as possible.
+		add_action( 'wp_body_open', array( $instance, 'add_bg_style_override' ), 999999 );
+	}
+
+	public function __construct( $lcp_data ) {
+		$this->lcp_data = $lcp_data;
+	}
+
+	public function preload_background_images() {
+		$selectors = array();
+
+		foreach ( $this->lcp_data as $lcp_data ) {
+			if ( in_array( $lcp_data['element'], $selectors, true ) ) {
+				// If we already printed the styling for this element, skip it.
+				continue;
+			}
+			$selectors[] = $lcp_data['element'];
+
+			$responsive_image_rules = $this->get_responsive_image_rules( $lcp_data );
+			$this->print_preload_links( $responsive_image_rules );
+		}
+	}
+
+	private function print_preload_links( $responsive_image_rules ) {
+		foreach ( $responsive_image_rules as $breakpoint ) {
+			$image_set = array();
+			foreach ( $breakpoint['image_set'] as $image ) {
+				$image_set[] = sprintf( '%s %sx', $image['url'], $image['dpr'] );
+			}
+
+			$image_set_string = implode( ', ', $image_set );
+
+			printf(
+				'<link rel="preload" href="%s" as="image" fetchpriority="high" media="%s" imagesrcset="%s" />' . PHP_EOL,
+				esc_url( Image_CDN_Core::cdn_url( $breakpoint['base_image'] ) ),
+				esc_attr( $breakpoint['media_query'] ),
+				esc_attr( $image_set_string )
+			);
+		}
+	}
+
+	public function add_bg_style_override() {
+		$selectors = array();
+
+		foreach ( $this->lcp_data as $lcp_data ) {
+			if ( in_array( $lcp_data['element'], $selectors, true ) ) {
+				// If we already printed the styling for this element, skip it.
+				continue;
+			}
+			$selectors[] = $lcp_data['element'];
+
+			$lcp_optimizer = new LCP_Optimization_Util( $lcp_data );
+			$image_url     = $lcp_optimizer->get_image_to_preload();
+			if ( empty( $image_url ) ) {
+				continue;
+			}
+
+			$styles                 = array();
+			$responsive_image_rules = $this->get_responsive_image_rules( $lcp_data );
+
+			// Add responsive image styling.
+			foreach ( $responsive_image_rules as $breakpoint ) {
+				$image_set = array();
+				foreach ( $breakpoint['image_set'] as $image ) {
+					$image_set[] = sprintf( 'url(%s) %sx', $image['url'], $image['dpr'] );
+				}
+
+				$image_set_string = implode( ', ', $image_set );
+
+				$styles[] = sprintf(
+					'@media %s { %s { background-image: url(%s) !important; background-image: -webkit-image-set(%s) !important; background-image: image-set(%s) !important; } }',
+					$breakpoint['media_query'],
+					$lcp_data['element'],
+					$breakpoint['base_image'],
+					$image_set_string,
+					$image_set_string
```

Comment:
Nitpick: we should be able to avoid issues like this with numbered placeholders. I think it would make things more readable, too.